client_plugin/drivers/shared/etcdops.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		log.WithFields(log.Fields{"nodeID": nodeID,
+			"error": err}).Error("Failted get NodeList from swarm manager")
+		return err
+	}
+	for _, n := range nodes {
+		if n.ManagerStatus != nil && n.ManagerStatus.Leader == true {
+			log.WithFields(
+				log.Fields{"leader ID": n.ID,
+					"manager ID": nodeID},
+			).Info("Found leader for swarm manager")
+
+			leaderAddr, err := checkManagerAddr(n)
+			if err != nil {
+				log.Errorf("Failed to checkManagerAddr")
+			} else {
+				joinEtcdCluster(addr, leaderAddr, nodeID)
+				return nil
+			}
+		}
+	}
+
+	err = fmt.Errorf("Failed to get leader for swarm manager %s", nodeID)
+	return err
+}
+
+// checkManagerAddr examine the state of a swarm node and return it's ManagerStatus.Addr
+// this function can only be called by manager node for now
+// when the docker API is upgraded to 1.25, this can be extended all the swarm node: Status.Addr
+func checkManagerAddr(node swarm.Node) (string, error) {
+	if node.Status.State != swarm.NodeStateReady {
+		log.WithFields(log.Fields{"nodeID": node.ID,
+			"state": node.Status.State}).Error("State of node is not ready ")
+		return "", fmt.Errorf("Node %s status %v", node.ID, node.Status.State)
+	}
+
+	// ManagerStatus.Addr is in the format of "host:port", return host only
+	s := strings.Split(node.ManagerStatus.Addr, ":")
+
+	return s[0], nil
+}
+
+// startEtcdCluster function is called by swarm leader to start a ETCD cluster
+func startEtcdCluster(nodeAddr string, nodeID string) error {
+	lines := []string{
+		"--name", nodeID,
+		"--advertise-client-urls", etcdScheme + nodeAddr + etcdClientPort,
+		"--initial-advertise-peer-urls", etcdScheme + nodeAddr + etcdPeerPort,
+		"--listen-client-urls", etcdScheme + etcdListenURL + etcdClientPort,
+		"--listen-peer-urls", etcdScheme + etcdListenURL + etcdPeerPort,
+		"--initial-cluster-token", etcdClusterToken,
+		"--initial-cluster", nodeID + "=" + etcdScheme + nodeAddr + etcdPeerPort,
+		"--initial-cluster-state", etcdClusterStateNew,
+	}
+	go etcdRoutine(lines)
+
+	return nil
+}
+
+// joinEtcdCluster function is called by a non-leader swarm manager to join a ETCD cluster
+func joinEtcdCluster(nodeAddr string, leaderAddr string, nodeID string) error {
+	endpoint := leaderAddr + etcdClientPort
+	cfg := etcdClient.Config{
+		Endpoints: []string{endpoint},
+	}
+
+	cli, err := etcdClient.New(cfg)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error": err},
+		).Error("Failed to create ETCD Client ")
+		return err
+	}
+	defer cli.Close()
+
+	// list all current ETCD members, check if this node is already added as a member
+	lresp, err := cli.MemberList(context.Background())
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error":       err,
+				"members len": len(lresp.Members)},
+		).Error("Failed to list member for ETCD")
+		return err
+	}
+
+	peerAddr := etcdScheme + nodeAddr + etcdPeerPort
```

Comment:
By functionality, peerAddr is appearing to be the address of this node. If so, can this be renamed to nodeAddr for clarity?